oap-server/server-receiver-plugin/skywalking-telegraf-receiver-plugin/src/main/java/org/apache/skywalking/oap/server/receiver/telegraf/provider/handler/TelegrafServiceHandler.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.skywalking.oap.server.telemetry.api.CounterMetrics;
+import org.apache.skywalking.oap.server.telemetry.api.HistogramMetrics;
+import org.apache.skywalking.oap.server.telemetry.api.MetricsCreator;
+import org.apache.skywalking.oap.server.telemetry.api.MetricsTag;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class TelegrafServiceHandler {
+
+    private final HistogramMetrics histogram;
+    private final CounterMetrics errorCounter;
+    private List<MetricConvert> metricConvert;
+
+    public TelegrafServiceHandler(ModuleManager moduleManager, MeterSystem meterSystem, List<TelegrafConfig> rules) {
+
+        this.metricConvert = rules.stream().map(r -> new MetricConvert(r, meterSystem)).collect(Collectors.toList());
+
+        final MetricsCreator metricsCreator = moduleManager.find(TelemetryModule.NAME)
+                                                           .provider()
+                                                           .getService(MetricsCreator.class);
+
+        histogram = metricsCreator.createHistogramMetric(
+                "telegraf_in_latency", "The process latency of telegraf data",
+                new MetricsTag.Keys("protocol"), new MetricsTag.Values("http")
+        );
+
+        errorCounter = metricsCreator.createCounter(
+                "telegraf_error_count", "The error number of telegraf analysis",
+                new MetricsTag.Keys("protocol"), new MetricsTag.Values("http")
+        );
+    }
+
+    /**
+    * Convert the TelegrafData object to meter {@link Sample}
+    **/
+    public List<Sample> convertTelegraf(TelegrafData telegrafData) {
+
+        List<Sample> sampleList = new ArrayList<>(Collections.emptyList());
+
+        Map<String, Object> fields = telegrafData.getFields();
+        String name = telegrafData.getName();
+        Map<String, String> tags = telegrafData.getTags();
+        ImmutableMap<String, String> immutableTags = ImmutableMap.copyOf(tags);
+        long timestamp = telegrafData.getTimestamp();
+
+        fields.forEach((key, value) -> {
+            Sample.SampleBuilder builder = Sample.builder();
+            if (value instanceof String) return;
+            Sample sample = builder.name(name + "_" + key)
+                    .timestamp(timestamp * 1000L)
+                    .value(((Number) value).doubleValue())
+                    .labels(immutableTags).build();
+
+            sampleList.add(sample);
+        });
+        return sampleList;
+
+    }
+
+    @Post("/telegraf")
+    public Commands collectData(String jsonInfo) {
+        TelegrafData telegrafData;
+        try (HistogramMetrics.Timer ignored = histogram.createTimer()) {
+            List<Sample> allSamples = new ArrayList<>();
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode node = mapper.readTree(jsonInfo);
```

Comment:
Can you rename the original `TelegrafData` class to `TelegrafDatum`, and define a class `TelegrafData` to contain a field `List<TelegrafDatum> metrics`, so you don't need to `readTree` here, and just declare the method `public Commands collectData(TelegrafData telegrafData) {`

Reply:
> Can you rename the original `TelegrafData` class to `TelegrafDatum`, and define a class `TelegrafData` to contain a field `List<TelegrafDatum> metrics`, so you don't need to `readTree` here, and just declare the method `public Commands collectData(TelegrafData telegrafData) {`

I change code according to your suggestion. Everything works fine.

This is `TelegrafData` class.
![image](https://user-images.githubusercontent.com/79000735/191666192-76d80f30-eed9-4569-ae26-3619578f0b98.png)

This is `public Commands collectData(TelegrafData telegrafData)` method.
![image](https://user-images.githubusercontent.com/79000735/191666960-bf395584-b0db-494f-b02e-3ed922c7863c.png)

Is this modification good? If there are still problems, what else do I need to fix? Thank you for your guidance!

Reply:
Not sure whether it's necessary that class `TelegrafData` should implement `RequestConverterFunction`, this is a standard POJO and the Armeria should be able to deserialize it correctly

Reply:
> Not sure whether it's necessary that class `TelegrafData` should implement `RequestConverterFunction`, this is a standard POJO and the Armeria should be able to deserialize it correctly

I set the class `TelegrafData` like this:
![image](https://user-images.githubusercontent.com/79000735/192081885-8811cae0-2432-4184-b197-c074d451c68f.png)

The method shows this **error**.
![image](https://user-images.githubusercontent.com/79000735/192082071-73e300a9-b1b5-4090-a395-a69ee0f17b47.png)

After I removed the code `@RequestConverter(TelegrafData.class)`, the oap-server can be run, but the class `TelegrafServiceHandler` can't work, and pops up **error** info:
![image](https://user-images.githubusercontent.com/79000735/192081965-2df5f991-e30c-4202-a709-c21b17785985.png)

According to the Armeria document, the POJO class should implement `RequestConverterFunction` that can be used in the `@RequestConverter` and `@RequestObject` annotations. You can check the section **Conversion between an HTTP message and a Java object** in this document [https://armeria.dev/docs/server-annotated-service/](url) or the part **Write a request converter** in this tutorials [https://armeria.dev/tutorials/rest/blog/implement-create/](url).

Maybe the class `TelegrafData` can be optimized, but I think class `TelegrafData `must implement `RequestConverterFunction` for converting JSON object to JAVA object.